### PR TITLE
misc: Add LULESH GPU tests

### PR DESCRIPTION
**Pull request by mkjost0**
This adds the LULESH tests, which currently run successfully, though they still use the gem5-resources directory to build the binary used as opposed to using a pre-built binary

Change-Id: I91c511fe92b7f9d11dfb027f435573f826bc6714

### CODE REVIEW

File: .github/workflows/gpu-tests.yaml
Comment (mattsinc) [inline .github/workflows/gpu-tests.yaml]: Same thing about j4 here as other ones

**Review by mattsinc — COMMENTED** [review]
Same question here about docker: I don't see where that is coming from.  Maybe the container on line 56?

**Review by abmerop — COMMENTED** [review]
Thanks for helping to move the GPU weekly tests out of the bash script!  This looks fine for GCN3.  One thing I would like though is to test the Vega ISA as well. I realize that will take more resources.  @mattsinc and I have been planning to discuss in detail if GCN3 can be deprecated. I would also like to see Vega as part of the ALL build, which would further reduce compilation for testing, I think (not sure if each yaml file is rebuilding).

I have many more detailed thoughts on moving to Vega but it is probably better for a discussions thread rather than this PR.

**Comment by mkjost0** [comment]
> Thanks for helping to move the GPU weekly tests out of the bash script! This looks fine for GCN3. One thing I would like though is to test the Vega ISA as well. I realize that will take more resources. @mattsinc and I have been planning to discuss in detail if GCN3 can be deprecated. I would also like to see Vega as part of the ALL build, which would further reduce compilation for testing, I think (not sure if each yaml file is rebuilding).
> 
> I have many more detailed thoughts on moving to Vega but it is probably better for a discussions thread rather than this PR.

Yes, this is something we can look into! I think if we do that, we can put it into a separate PR so we can update all the GPU tests to use Vega at the same time after we ensure it works locally first, unless you have other thoughts on that.

File: .github/workflows/gpu-tests.yaml
Comment (mattsinc) [inline .github/workflows/gpu-tests.yaml]: @mkjost0 similar to other apps, how did you determine this timeout?  My recollection is that LULESH sometimes runs longer than this when running the default 10 iterations -- maybe @abmerop remembers differently though...

**Review by mattsinc — DISMISSED** [review]
Only one question, which may be moot if you verified.

File: .github/workflows/gpu-tests.yaml
Comment (abmerop) [reply .github/workflows/gpu-tests.yaml]: A full run takes around 15 hours on my machine.  I usually exit after the first iteration when testing now though. I am not sure how many this test is trying to run

Comment (mattsinc) [reply .github/workflows/gpu-tests.yaml]: Since there are no input args to the contrary, I believe it is running the full 10 iterations: https://gem5.googlesource.com/public/gem5-resources/+/refs/heads/develop/src/gpu/lulesh/src/lulesh.hip.cc#5507

Your experience squares with mine in terms of runtime.

Comment (Harshil2107) [reply .github/workflows/gpu-tests.yaml]: Melissa is no longer working on the project. I don't know how she got the timeout number. How long would you recommend setting the timeout?

Comment (mattsinc) [reply .github/workflows/gpu-tests.yaml]: I was not aware Melissa wasn't working on it anymore, sorry.

Have you tried running LULESH in your setup?  If so, how long does it take?  I would set the timeout accordingly.

Comment (abmerop) [reply .github/workflows/gpu-tests.yaml]: 24 hours should be sufficient... But I would like to discuss how redundant the run for this app is.  We don't gain much testing the default 10 iterations in this app vs. one, for example.  And with one it would take ~2 hours.

Comment (mattsinc) [reply .github/workflows/gpu-tests.yaml]: There is an input arg to control the number of iterations, so making this change is straightforward.

In terms of why multiple -- I think there is value in at least running 2 iterations to make sure inter-kernel behavior works.  But I agree 10 is a lot.

File: src/arch/arm/regs/misc.cc
Comment (mattsinc) [inline src/arch/arm/regs/misc.cc]: I'm not sure what these ARM changes have to do with the LULESH benchmark?

File: .github/workflows/gpu-tests.yaml
Comment (mattsinc) [inline .github/workflows/gpu-tests.yaml]: @BobbyRBruce : per @abmerop and I's conversation on the previous version, please use "0.01 2" as the input args and then reduce the timeout to 8 hours.

**Review by mattsinc — CHANGES_REQUESTED** [review]

File: .github/workflows/gpu-tests.yaml
Comment (BobbyRBruce) [reply .github/workflows/gpu-tests.yaml]: When you say passing the input args, do you mean like this?: ` build/GCN3_X86/gem5.opt configs/example/apu_se.py -n3 --mem-size=8GB --reg-alloc-policy=dynamic --benchmark-root="lulesh" -c lulesh 0.01 2` ?

File: src/arch/arm/regs/misc.cc
Comment (BobbyRBruce) [reply src/arch/arm/regs/misc.cc]: I'm going to separate this out, but TL;DR I needed to add this here for now to commit. "misc.cc" breaks the line-length rules and my pre-commit was complaining. I'll let you know when it's gone

File: .github/workflows/gpu-tests.yaml
Comment (mattsinc) [reply .github/workflows/gpu-tests.yaml]: Yes, that looks correct

Comment (BobbyRBruce) [reply .github/workflows/gpu-tests.yaml]: Done.

**Review by mattsinc — DISMISSED** [review]
thanks!  Please let @abmerop confirm before you merge in though.

File: .github/workflows/gpu-tests.yaml
Comment (mattsinc) [inline .github/workflows/gpu-tests.yaml:65]: Actually, now that I look at this, did you test this?  I am surprised line 65 is "lulesh 0.01 2" instead of "0.01 2" -- maybe this is something with the new YAML format I'm not getting?

Comment (abmerop) [reply .github/workflows/gpu-tests.yaml:65]: It looks right to me.  I am using the left/right view on github so the original "lulesh" (no parameters) command was moved to a new line along with adding the parameters "0.01 2".  In my opinion the other view is more confusing

**Review by abmerop — DISMISSED** [review]

**Comment by abmerop** [comment]
> thanks! Please let @abmerop confirm before you merge in though.

done

File: .github/workflows/gpu-tests.yaml
Comment (mattsinc) [reply .github/workflows/gpu-tests.yaml:65]: Weird, just must be my browser's view then.

File: src/arch/arm/regs/misc.cc
Comment (abmerop) [reply src/arch/arm/regs/misc.cc]: I think I have seen this when merging develop into my private branch. Thanks for fixing this!

Comment (BobbyRBruce) [reply src/arch/arm/regs/misc.cc]: I messed up the fix in here. I'll wait for my real fix for this in #459 to be merged to develop then fix it all.

**Comment by BobbyRBruce** [comment]
FYI, when this is good to go (soon, I promise), i'll do a merge commit with this one so it'll be added to develop as one commit.

Just giving a heads up before people start to notice all the touch-up commits i've been doing.

**Review by mattsinc — COMMENTED** [review]
Sorry I could not review this faster @BobbyRBruce , I was traveling the last ~5 days.  I think the final version looks good, thanks!

**Comment by Harshil2107** [comment]
Hi @mattsinc,

The Lulesh tests are encountering failures due to an apparent issue with the run command. The current command reads as follows: `build/GCN3_X86/gem5.opt configs/example/apu_se.py -n3 --mem-size=8GB --reg-alloc-policy=dynamic --benchmark-root=lulesh -c lulesh 0.01 2`. Upon attempting to run the lulesh test locally, I received an error indicating that 0.01 and 2 are invalid arguments. Despite scanning through the code, I couldn't identify any section within apu_se.py that accepts these arguments. Could you provide guidance on this matter?

**Comment by abmerop** [comment]
Hi @Harshil2107 ,  I was actually just looking at this and was wondering how this command was even working. It seems it's not.  I will post a comment on what I suspect is the issue.

File: .github/workflows/gpu-tests.yaml
Comment (abmerop) [inline .github/workflows/gpu-tests.yaml:65]: I think this needs to be something like:

`build/GCN3_X86/gem5.opt configs/example/apu_se.py -n3 --mem-size=8GB --reg-alloc-policy=dynamic --benchmark-root="lulesh" -c lulesh --options="0.01 2"`

I suspect otherwise the 0.01 and 2 are being passed as positional arguments to the apu_se.py script

Comment (mattsinc) [reply .github/workflows/gpu-tests.yaml:65]: Yes @abmerop is correct.  Those inputs are arguments to to the benchmark, not to apu_se.py.

**Pull request merged**
